lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,382 @@
+  int _compareExpansionKeys(String keyA, String keyB) {
+    var baseLengthA = keyA.contains('*') ? keyA.indexOf('*') + 1 : keyA.length;
+    var baseLengthB = keyB.contains('*') ? keyB.indexOf('*') + 1 : keyB.length;
+    if (baseLengthA > baseLengthB) return -1;
+    if (baseLengthB > baseLengthA) return 1;
+    if (!keyA.contains("*")) return 1;
+    if (!keyB.contains("*")) return -1;
+    if (keyA.length > keyB.length) return -1;
+    if (keyB.length > keyA.length) return 1;
+    return 0;
+  }
+
+  /// Returns a file path for `subpath`, as resolved in the `exports` object.
+  ///
+  /// Verifies the file exists relative to `packageRoot`. Instances of `*` will
+  /// be replaced with `patternMatch`.
+  ///
+  /// `subpath` and `packageRoot` are native paths, and `patternMatch` is a URL
+  /// path.
+  ///
+  /// Implementation of `PACKAGE_TARGET_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _packageTargetResolve(
+      String? subpath, Object exports, String packageRoot,
+      [String? patternMatch]) {
+    switch (exports) {
+      case String string when !string.startsWith('./'):
+        throw "Export '$string' must be a path relative to the package root at '$packageRoot'.";
+      case String string when patternMatch != null:
+        var replaced = p.fromUri(string.replaceFirst('*', patternMatch));
+        var path = p.normalize(p.join(packageRoot, replaced));
+        return fileExists(path) ? path : null;
+      case String string:
+        return p.join(packageRoot, p.fromUri(string));
+      case Map<String, dynamic> map:
+        for (var (key, value) in map.pairs) {
+          if (!const {'sass', 'style', 'default'}.contains(key)) continue;
+          if (value == null) continue;
+          if (_packageTargetResolve(
+                  subpath, value as Object, packageRoot, patternMatch)
+              case var result?) {
+            return result;
+          }
+        }
+        return null;
+
+      case []:
+        return null;
+
+      case List<dynamic> array:
+        for (var value in array) {
+          if (value == null) continue;
+          if (_packageTargetResolve(
+                  subpath, value as Object, packageRoot, patternMatch)
+              case var result?) {
+            return result;
+          }
+        }
+
+        return null;
+
+      default:
+        throw "Invalid 'exports' value $exports in "
+            "${p.join(packageRoot, 'package.json')}.";
+    }
+  }
+
+  /// Returns a path to a package's export without a subpath.
+  Object? _getMainExport(Object exports) {
+    return switch (exports) {
+      String string => string,
+      List<String> list => list,
+      Map<String, dynamic> map
+          when !map.keys.any((key) => key.startsWith('.')) =>
+        map,
+      Map<String, dynamic> map when map.containsKey('.') && map['.'] != null =>
```

Comment:
You can make this a map pattern: `<String, dynamic>{'.': var export?} => export`